pkg/controller/job/job_controller.go

```diff
@@ -1290,15 +1296,26 @@ func getStatus(job *batch.Job, pods []*v1.Pod, uncounted *uncountedTerminatedPod
 		return p.Status.Phase == v1.PodSucceeded
 	}))
 	failed += int32(countValidPodsWithFilter(job, pods, uncounted.Failed(), expectedRmFinalizers, func(p *v1.Pod) bool {
-		if p.Status.Phase == v1.PodFailed {
-			return true
+		isFailed := p.Status.Phase == v1.PodFailed
```

Comment:
it's hard to tell that this is preserving the behavior when the feature is disabled.

I suggest you query the feature gate here, and if it's disabled, run the old code, otherwise, run the new code.

It's ok to duplicate code when we are adding a new gated feature.